### ISSUE TITLE
Guess if the presence or the absence of a feature helps to be in a specific class 

### ISSUE THREAD

**Issue by pommedeterresautee** [tqchen/xgboost#166]
Each predictor improves the purity of the branch it is on. The purity is measured by the gain.

In a binary classification, some features need to be present in an observation to predict that the observation will be classified as 1 and some other predictors need to be absent to get the exact same result.

My question is how, **based on the model text dump of a binary classification**, can I say that the presence of a feature A helps to predict that an observation will be classified as 1 and another feature B absence helps to predict that an observation will be classified as 1.

model may look like that
```R
> xgb.dump(bst, with.stats = T)
 [1] "booster[0]"                                                     
 [2] "0:[f28<1.00001] yes=1,no=2,missing=2,gain=4000.53,cover=1628.25"
 [3] "1:[f108<1.00001] yes=3,no=4,missing=4,gain=198.174,cover=703.75"
 [4] "3:leaf=1.85965,cover=13.25"                                     
 [5] "4:leaf=-1.94071,cover=690.5"                                    
 [6] "2:[f55<1.00001] yes=5,no=6,missing=6,gain=1158.21,cover=924.5"  
 [7] "5:leaf=-1.70044,cover=112.5"                                    
 [8] "6:leaf=1.71218,cover=812"                                       
 [9] "booster[1]"                                                     
[10] "0:[f59<1.00001] yes=1,no=2,missing=2,gain=832.545,cover=788.852"
[11] "1:leaf=-6.23624,cover=20.4624"                                  
[12] "2:[f28<1.00001] yes=3,no=4,missing=4,gain=569.725,cover=768.39" 
[13] "3:leaf=-0.96853,cover=309.453"                                  
[14] "4:leaf=0.784718,cover=458.937"   
```

**Comment by pommedeterresautee** [comment]
In my understanding, I think I should analyze each path beginning from the leaf. But I am not sure. Moreover can someone confirms the meaning of <1.0001 for a dummy feature?

**Comment by pommedeterresautee** [comment]
The idea is to bring more information in the feature importance analysis.

**Comment by pommedeterresautee** [comment]
@tqchen if it is possible, if it makes sense, and if it is easy to do, it would be really good to add this information to the model dump (otherwise computing every path for every tree may be performance costly)

**Comment by tqchen** [comment]
I do not get your question.. Normally, it is hard to calculate one feature's contribution to specific class, in binary classification case, a good positive feature correlated with positive class also means it is good to discriminate the negative ones.

The gain is calculated by the gain in terms of loss function, i.e. how good the split is to discriminate the pos/negs for classification case.

For indicator feature, it is better to dump with feature map. The f28 < 1.00001  means when f28=1 it goes to left branch. Note that missing direction is right branch, so when f28 is not presented in the sparse matrix, it goes to right branch

**Comment by pommedeterresautee** [comment]
You are right, I was not very clear in my mind and in my question.

I understand your point, a good feature discriminate both classes in a binary classification.

What I am interested in, is the co-occurrence of a feature and a specific class.

Let's say I have a dataset with just one categorical variable, "color", and an outcome column Y which encodes a binary classification (1 or 0). Inside the categorical columns there are 100 different values, red, blue, green, ...

The outcome column is such that it is 1 when the color is NOT red and 0 when it is red.

Therefore P(color=Red / Y = 0) = 1 

So I will one hot encode my variable and get 100 new binary column, one per color. Xgboost should easily find that the most discriminating feature is the binary column encoding presence of red color. So the gain will be super high for this feature. But co-occurence of Y=1 and color Red = 1 will be zero.

This information is very valuable in feature importance analysis. Is there a way to extract it from existing information? 

Today I am counting the co-occurrence of the most important feature directly on the dataset. But I think it would be very useful for everyone if I can include such information directly in the result of feature importance function of Xgboost.

Kind regards,
Michaël

**Comment by tqchen** [comment]
We cannot get the statistics information from the current dump. The reason is that the tree-solver is designed for solving general tasks, including regression, classification, rank etc. And to do that we use a common protocol of loss reduction(gain). So label information is simply not available in tree learning(the effect pass in through gradient statistics). 

Such protocol simplifies the design and make s the interface minimum and general for all of tasks, and I think it is good to keep it so. The conditionally statistics could be obtained by a explicit counting, if the user is interested in doing so.

**Comment by pommedeterresautee** [comment]
ok it makes sense. Tks
